import os
import json
from PySide6.QtWidgets import (
    QWidget, QPushButton, QLabel, QMessageBox, QFileDialog, QTextBrowser,
    QGridLayout, QSizePolicy, QDialog, QVBoxLayout,
    QHBoxLayout, QScrollArea
)
from PySide6.QtCore import Qt, QTimer, QCoreApplication, QIODevice
from PySide6.QtGui import QTextOption, QColor, QPalette  # å¯¼å…¥ QColor å’Œ QPalette

# å¯¼å…¥æ ¸å¿ƒæ¨¡å— (ä½¿ç”¨ç»å¯¹å¯¼å…¥)
from ui_utils import BaseTabWidget, WorkerThread

# ä¿æŒå¯¹åŒçº§æ¨¡å—çš„ç›¸å¯¹å¯¼å…¥
from gemini.widgets.category_dialog import CategorySelectionDialog

# ç¡®ä¿å¯¼å…¥ Service (é€šè¿‡ services/__init__.py æ¡¥æ¥å¯¼å…¥)
from services import FileManager


class BatchTabWidget(BaseTabWidget):
    def __init__(self, qingshilu_service, parent=None):
        super().__init__("batch_tab.ui", parent)

        # ä¿®æ­£ 1: ç¡®ä¿ BatchTabWidget è‡ªèº«èƒ½å¤Ÿæ°´å¹³å’Œå‚ç›´æ‹‰ä¼¸
        self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)

        # ğŸŒŸ æ–°å¢çŠ¶æ€å˜é‡ï¼šè®°å½•å½“å‰æ˜¯å¦å¤„äºç­›é€‰çŠ¶æ€ ğŸŒŸ
        self.is_filtered = False

        # é€šè¿‡ä¾èµ–æ³¨å…¥è·å– Service å®ä¾‹
        self.qingshilu_service = qingshilu_service
        self.file_manager = FileManager(qingshilu_service)
        self.worker = None

        # ğŸŒŸ å…³é”®ï¼šè°ƒç”¨ä¸»é¢˜é€‚åº”é€»è¾‘ ğŸŒŸ
        self.apply_theme_adaptation()

        self.connect_signals()

        # å…³é”®ä¿®æ­£: è°ƒç”¨æ‹‰ä¼¸è®¾ç½®æ–¹æ³•
        self._ensure_vertical_stretch()

    def apply_theme_adaptation(self):
        """
        æ ¹æ®ç³»ç»Ÿä¸»é¢˜ï¼Œè®¾ç½®æµ…è‰²æ¨¡å¼ä¸‹çš„èƒŒæ™¯è‰²ä¸ºç±³ç™½è‰² (#FFFFF0)ï¼Œ
        æ·±è‰²æ¨¡å¼ä¸‹ä½¿ç”¨ Qt çš„ç³»ç»Ÿé»˜è®¤èƒŒæ™¯è‰²ï¼Œå®ç°è‡ªé€‚åº”ã€‚
        """
        palette = self.palette()

        # è·å–ç³»ç»Ÿçª—å£é¢œè‰²ä½œä¸ºåˆ¤æ–­ä¾æ®
        window_color = palette.color(QPalette.Window)

        # åˆ¤æ–­æ˜¯å¦ä¸ºæ·±è‰²æ¨¡å¼ï¼šå¦‚æœçª—å£é¢œè‰²çš„äº®åº¦è¾ƒä½ï¼ˆä¾‹å¦‚ï¼ŒR, G, Béƒ½å°äº128ï¼‰ï¼Œåˆ™è§†ä¸ºæ·±è‰²æ¨¡å¼ã€‚
        # æ³¨æ„ï¼šè¿™æ˜¯ä¸€ä¸ªåŸºäºé¢œè‰²çš„å¯å‘å¼åˆ¤æ–­ï¼Œå¯èƒ½ä¸å®Œç¾ï¼Œä½†åœ¨QSSä¸æ”¯æŒåª’ä½“æŸ¥è¯¢æ—¶æ˜¯å¸¸ç”¨æ–¹æ³•ã€‚
        is_dark_mode = window_color.red() < 128 and window_color.green() < 128 and window_color.blue() < 128

        if not is_dark_mode:
            # â˜€ï¸ æµ…è‰²æ¨¡å¼ï¼šå¼ºåˆ¶ä½¿ç”¨ç±³ç™½è‰² (#FFFFF0)
            custom_color = QColor("#FFFFF0")

            # è®¾ç½®æ•´ä¸ª Tab é¡¶å±‚ Widget çš„èƒŒæ™¯
            palette.setColor(QPalette.Window, custom_color)

            # ç¡®ä¿åº”ç”¨äº†èƒŒæ™¯è‰²
            self.setAutoFillBackground(True)
            self.setPalette(palette)

            # é‡æ–°è®¾ç½® QTextBrowser çš„æŠ¤çœ¼è‰²å¸¸é‡ï¼Œç¡®ä¿æ¸²æŸ“æ—¶èƒ½å¤Ÿä½¿ç”¨
            self.EYE_CARE_COLOR = "#FFFFF0"
        else:
            # ğŸŒ™ æ·±è‰²æ¨¡å¼ï¼šä¿æŒé»˜è®¤ï¼Œè·Ÿéšç³»ç»Ÿä¸»é¢˜
            self.setAutoFillBackground(False)

            # åœ¨æ·±è‰²æ¨¡å¼ä¸‹ï¼Œæˆ‘ä»¬å¸Œæœ›æ¡æ–‡å¡ç‰‡ä½¿ç”¨ç³»ç»ŸèƒŒæ™¯è‰²
            dark_mode_bg_color = palette.color(QPalette.Window).name()
            self.EYE_CARE_COLOR = dark_mode_bg_color  # æ·±è‰²æ¨¡å¼ä¸‹ä½¿ç”¨ç³»ç»Ÿé¢œè‰²

    def _ensure_vertical_stretch(self):
        """ç¡®ä¿ QScrollArea (ç´¢å¼• 1) å æ® BatchTabWidget å‚ç›´æ‹‰ä¼¸ç©ºé—´ã€‚"""
        main_layout = self.layout()

        if not isinstance(main_layout, QVBoxLayout):
            return

        # å¸ƒå±€ç»“æ„ç¡®è®¤ (åŸºäº batch_tab.ui):
        # ç´¢å¼• 0: selectBatchFileButton (Stretch=0)
        # ç´¢å¼• 1: batchScrollArea (Stretch=1)
        # ç´¢å¼• 2: batchButtonLayout (Stretch=0)
        # ç´¢å¼• 3: notificationLabel (Stretch=0)

        if main_layout.count() >= 4:
            # èµ‹äºˆ QScrollArea (ç´¢å¼• 1) æ‰€æœ‰å‚ç›´æ‹‰ä¼¸ç©ºé—´
            main_layout.setStretch(1, 1)

            # ç¡®ä¿æŒ‰é’®å’Œæ ‡ç­¾ä¸æ‹‰ä¼¸ï¼Œä¿æŒç´§å‡‘
            main_layout.setStretch(0, 0)
            main_layout.setStretch(2, 0)
            main_layout.setStretch(3, 0)
        else:
            print("Warning: BatchTabWidget layout count is unexpected. Vertical stretch skipped.")

    def connect_signals(self):
        self.selectBatchFileButton = self.findChild(QPushButton, "selectBatchFileButton")
        self.processBatchButton = self.findChild(QPushButton, "processBatchButton")
        self.saveBatchButton = self.findChild(QPushButton, "saveBatchButton")
        # ğŸŒŸ æ–°å¢ï¼šè¿æ¥ç­›é€‰æŒ‰é’® ğŸŒŸ
        self.filterUnclassifiedButton = self.findChild(QPushButton, "filterUnclassifiedButton")

        self.batchContents = self.findChild(QWidget, "batchContents")

        self.notificationLabel = self.findChild(QLabel, "notificationLabel")
        if self.notificationLabel:
            self.notificationLabel.hide()

        if self.selectBatchFileButton:
            self.selectBatchFileButton.clicked.connect(self._handle_select_batch_files_controller)
        if self.processBatchButton:
            self.processBatchButton.clicked.connect(self._start_process_batch_worker)
        if self.saveBatchButton:
            self.saveBatchButton.clicked.connect(self._handle_save_batch_results)
        # ğŸŒŸ è¿æ¥ç­›é€‰æŒ‰é’®çš„ä¿¡å· ğŸŒŸ
        if self.filterUnclassifiedButton:
            self.filterUnclassifiedButton.clicked.connect(self._toggle_filter_unclassified)

    def _toggle_filter_unclassified(self):
        """
        åˆ‡æ¢ç­›é€‰çŠ¶æ€ï¼šæ˜¾ç¤ºæ‰€æœ‰æ¡æ–‡ æˆ– åªæ˜¾ç¤ºæœªåˆ†ç±»æ¡æ–‡ã€‚
        """
        if self.is_filtered:
            # åˆ‡æ¢å›æ˜¾ç¤ºå…¨éƒ¨
            self.is_filtered = False
            self.filterUnclassifiedButton.setText("ç­›é€‰æœªåˆ†ç±»")
            self.filterUnclassifiedButton.setStyleSheet("background-color: #3f689f; color: white;")
            results_to_render = self.file_manager.get_batch_articles()
            self.show_notification(f"å·²æ˜¾ç¤ºå…¨éƒ¨ {len(results_to_render)} æ¡æ¡æ–‡ã€‚")
        else:
            # åˆ‡æ¢åˆ°ç­›é€‰æ¨¡å¼
            self.is_filtered = True
            self.filterUnclassifiedButton.setText("æ˜¾ç¤ºå…¨éƒ¨")
            self.filterUnclassifiedButton.setStyleSheet("background-color: #ffc107; color: black;")  # é†’ç›®é¢œè‰²

            all_articles = self.file_manager.get_batch_articles()
            # ç­›é€‰å‡ºåˆ†ç±»é”®ä¸º None æˆ–ç©ºå­—ç¬¦ä¸²çš„æ¡æ–‡
            results_to_render = [
                a for a in all_articles
                if a.get('classification_key') is None or a.get('classification_key') == ''
            ]
            self.show_notification(f"å·²ç­›é€‰å‡º {len(results_to_render)} æ¡æœªåˆ†ç±»æ¡æ–‡ã€‚")

        self._render_batch_results(results_to_render)

    def _handle_select_batch_files_controller(self):
        """å¤„ç†é€‰æ‹©æ‰¹é‡æ–‡ä»¶æŒ‰é’®çš„ç‚¹å‡»äº‹ä»¶ (Controller èŒè´£)"""
        files = self.file_manager.select_batch_files(self)

        if files:
            # ğŸŒŸ é‡ç½®ç­›é€‰çŠ¶æ€ ğŸŒŸ
            self.is_filtered = False
            self.filterUnclassifiedButton.setText("ç­›é€‰æœªåˆ†ç±»")
            self.filterUnclassifiedButton.setStyleSheet("background-color: #3f689f; color: white;")

            batch_list_widget = self.findChild(QWidget, "batchContents")
            batch_list_layout = batch_list_widget.layout()

            if batch_list_layout:
                # æ¸…ç©ºæ—§æ–‡ä»¶åˆ—è¡¨
                while batch_list_layout.count():
                    child = batch_list_layout.takeAt(0)
                    if child and child.widget():
                        child.widget().deleteLater()

                # Controller è´Ÿè´£ UI çš„æ›´æ–°
                for file in files:
                    file_label = QLabel(os.path.basename(file))
                    file_label.setToolTip(file)
                    batch_list_layout.addWidget(file_label)

                batch_list_layout.addStretch()
                QMessageBox.information(self, "æç¤º", f"å·²é€‰æ‹©äº† {len(files)} ä¸ªæ–‡ä»¶ã€‚")

    def _start_process_batch_worker(self):
        """å¯åŠ¨å¼‚æ­¥æ‰¹é‡å¤„ç†çº¿ç¨‹"""
        if not self.file_manager.get_selected_files():
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·å…ˆé€‰æ‹©éœ€è¦å¤„ç†çš„æ–‡ä»¶ã€‚")
            return

        self.processBatchButton.setEnabled(False)
        self.processBatchButton.setText("æ‰¹é‡å¤„ç†ä¸­...")
        # ğŸŒŸ ç¦ç”¨ç­›é€‰æŒ‰é’® ğŸŒŸ
        self.filterUnclassifiedButton.setEnabled(False)

        self.worker = WorkerThread(self.file_manager.process_files)

        self.worker.result_signal.connect(self._on_batch_success)
        self.worker.error_signal.connect(self._on_batch_error)

        self.worker.start()

    def _on_batch_success(self, message):
        """æ‰¹é‡å¤„ç†å®Œæˆååœ¨ä¸»çº¿ç¨‹ä¸­æ‰§è¡Œï¼Œå¹¶æ˜¾ç¤ºç»“æœæ¦‚è§ˆ"""
        self.processBatchButton.setEnabled(True)
        self.processBatchButton.setText("æ‰¹é‡åˆ†æ")
        # ğŸŒŸ å¯ç”¨ç­›é€‰æŒ‰é’® ğŸŒŸ
        self.filterUnclassifiedButton.setEnabled(True)

        # å§‹ç»ˆä½¿ç”¨å®Œæ•´çš„æ–‡ç« åˆ—è¡¨è¿›è¡Œæ¸²æŸ“ï¼Œè®©ç”¨æˆ·å†³å®šæ˜¯å¦ç­›é€‰
        results = self.file_manager.get_batch_articles()

        # æ‰¹é‡å¤„ç†æˆåŠŸåï¼Œé‡ç½®ç­›é€‰çŠ¶æ€ï¼Œæ˜¾ç¤ºå…¨éƒ¨
        self.is_filtered = False
        self.filterUnclassifiedButton.setText("ç­›é€‰æœªåˆ†ç±»")
        self.filterUnclassifiedButton.setStyleSheet("background-color: #3f689f; color: white;")

        self._render_batch_results(results)

        QMessageBox.information(self, "æ‰¹é‡å®Œæˆ", message)

    def _on_batch_error(self, error_message):
        """æ‰¹é‡å¤„ç†å¤±è´¥ååœ¨ä¸»çº¿ç¨‹ä¸­æ‰§è¡Œ"""
        self.processBatchButton.setEnabled(True)
        self.processBatchButton.setText("æ‰¹é‡åˆ†æ")
        # ğŸŒŸ å¯ç”¨ç­›é€‰æŒ‰é’® ğŸŒŸ
        self.filterUnclassifiedButton.setEnabled(True)
        QMessageBox.critical(self, "é”™è¯¯", f"æ‰¹é‡å¤„ç†è¿‡ç¨‹ä¸­å‘ç”Ÿé”™è¯¯ï¼š\n{error_message}")

    def show_notification(self, message: str, is_error: bool = False):
        """æ˜¾ç¤ºä¸´æ—¶çš„åº•éƒ¨é€šçŸ¥ï¼Œ3ç§’åè‡ªåŠ¨éšè—"""
        if not self.notificationLabel:
            return

        style = "padding: 5px; border-radius: 3px; font-weight: bold;"

        if is_error:
            style += "background-color: #f8d7da; color: #721c24;"
        else:
            style += "background-color: #d4edda; color: #155724;"

        self.notificationLabel.setText(message)
        self.notificationLabel.setStyleSheet(style)
        self.notificationLabel.show()

        QTimer.singleShot(3000, self.notificationLabel.hide)

    def _render_batch_results(self, results):
        """å°†æ‰¹é‡å¤„ç†çš„æ¡æ–‡ç»“æœæ¸²æŸ“åˆ° UI ä¸­"""
        if not self.batchContents: return

        layout = self.batchContents.layout()
        if not layout:
            layout = QVBoxLayout(self.batchContents)
            self.batchContents.setLayout(layout)

        # æ¸…ç©ºæ—§å†…å®¹
        while layout.count():
            child = layout.takeAt(0)
            if child and child.widget():
                child.widget().deleteLater()

        # ğŸŒŸ ä½¿ç”¨åœ¨ apply_theme_adaptation ä¸­è®¾ç½®çš„è‡ªé€‚åº”é¢œè‰² ğŸŒŸ
        # æµ…è‰²æ¨¡å¼ä¸‹æ˜¯ #FFFFF0ï¼Œæ·±è‰²æ¨¡å¼ä¸‹æ˜¯ç³»ç»Ÿé»˜è®¤æ·±è‰²
        ARTICLE_CARD_BG = getattr(self, 'EYE_CARE_COLOR', "#FFFFF0")  # è·å–è‡ªé€‚åº”é¢œè‰²ï¼Œé»˜è®¤ä»ä¸ºç±³ç™½è‰²

        # æ¸²æŸ“æ–°ç»“æœ
        for result in results:
            if 'error' in result:
                # æ¸²æŸ“é”™è¯¯ä¿¡æ¯
                article_id = result.get('article_id', 'æœªçŸ¥é”™è¯¯')
                label = QLabel(f"é”™è¯¯: {article_id}\nä¿¡æ¯: {result['error']}")
                label.setStyleSheet(
                    "color: red; font-weight: bold; padding: 5px; border: 1px solid red; background-color: #FFE0E0;")
                label.setSizePolicy(QSizePolicy(QSizePolicy.Expanding, QSizePolicy.Preferred))
                layout.addWidget(label)
                continue

            # æ¸²æŸ“æ¡æ–‡åˆ†æç»“æœ
            article_id = result['article_id']
            category_key = result.get('classification_key')

            article_group = QWidget()
            sizePolicy = QSizePolicy(QSizePolicy.Expanding, QSizePolicy.Preferred)
            article_group.setSizePolicy(sizePolicy)

            group_layout = QGridLayout(article_group)

            # ğŸŒŸ é‡ç‚¹ä¿®æ”¹ 2: è®¾ç½® article_group çš„èƒŒæ™¯ä¸ºè‡ªé€‚åº”é¢œè‰² ğŸŒŸ
            article_group.setStyleSheet(
                f"""
                QWidget {{
                    background-color: {ARTICLE_CARD_BG}; 
                    border: 1px solid #ddd; 
                    margin: 5px; 
                    padding: 5px;
                }}
                """
            )
            group_layout.setColumnStretch(0, 1)

            # 1. ID å’ŒåŸæ–‡
            id_label = QLabel(f"<b>ID: {article_id}</b>")
            text_browser = QTextBrowser()
            text_browser.setWordWrapMode(QTextOption.WrapAtWordBoundaryOrAnywhere)
            original_text = result.get('originalText', 'åŸæ–‡å†…å®¹ç¼ºå¤±')
            text_browser.setText(original_text)
            text_browser.setMinimumHeight(150)
            text_browser.setFrameShape(QTextBrowser.NoFrame)

            # ğŸŒŸ é‡ç‚¹ä¿®æ”¹ 3: ç¡®ä¿ QTextBrowser èƒŒæ™¯ä¸ article_group èƒŒæ™¯ä¸€è‡´ï¼Œå¹¶è®¾ç½®å­—ä½“ ğŸŒŸ
            # æµ…è‰²æ¨¡å¼ä¸‹å­—ä½“ä¸ºé»‘è‰²ï¼Œæ·±è‰²æ¨¡å¼ä¸‹å­—ä½“åº”ä¸ºç™½è‰²
            text_color = "black" if ARTICLE_CARD_BG == "#FFFFF0" else "white"

            text_browser.setStyleSheet(
                f"""
                QTextBrowser {{
                    background-color: {ARTICLE_CARD_BG}; /* è‡ªé€‚åº”èƒŒæ™¯è‰² */
                    color: {text_color}; /* è‡ªé€‚åº”å­—ä½“é¢œè‰² */
                    font-size: 20pt; 
                    line-height: 1.5;
                }}
                """
            )

            # 2. æ¨èåˆ†ç±»
            recommendations = result['analysis'].get('recommendations', [])
            rec_container = QWidget()
            rec_layout = QHBoxLayout(rec_container)
            rec_layout.setContentsMargins(0, 0, 0, 0)
            rec_layout.setSpacing(5)

            # ğŸŒŸ ç¡®ä¿æ¨èå®¹å™¨èƒŒæ™¯ä¸æ¡æ–‡å¡ç‰‡èƒŒæ™¯ä¸€è‡´ ğŸŒŸ
            rec_container.setStyleSheet(f"background-color: {ARTICLE_CARD_BG};")

            if recommendations:
                for i, rec in enumerate(recommendations[:3]):
                    recommendation_key = rec['category']
                    display_text = f"æ¨è{i + 1}: {recommendation_key}"
                    rec_button = QPushButton(display_text)
                    rec_button.article_id = article_id
                    rec_button.recommendation_key = recommendation_key
                    # æŒ‰é’®æ ·å¼ä¿æŒä¸å˜ï¼Œä½†å¯ä»¥è€ƒè™‘åœ¨æ·±è‰²æ¨¡å¼ä¸‹è°ƒæ•´é¢œè‰²ä»¥å¢åŠ å¯¹æ¯”åº¦
                    rec_button.setStyleSheet(
                        "QPushButton { color: #1a3b2e; background-color: #e0f7e0; border: 1px solid #a8dfa8; padding: 3px 6px; } QPushButton:hover { background-color: #c0f0c0; }")
                    rec_button.setToolTip(f"ç‚¹å‡»å³å¯é‡‡ç”¨æ¨è{i + 1}ï¼š{recommendation_key}å¹¶ä¿å­˜")
                    rec_button.clicked.connect(self._handle_accept_recommendation)
                    rec_layout.addWidget(rec_button)

                rec_layout.addStretch()
                rec_widget = rec_container
            else:
                rec_label = QLabel("æ¨è: æ— ")
                rec_label.setMinimumWidth(150)
                # ç¡®ä¿æ ‡ç­¾é¢œè‰²åœ¨æ·±è‰²æ¨¡å¼ä¸‹å¯è§
                rec_label.setStyleSheet(f"color: {text_color};")
                rec_widget = rec_label

            # 3. å½“å‰åˆ†ç±»çŠ¶æ€
            current_cat_text = category_key if category_key else "æœªåˆ†ç±»"
            current_cat_label = QLabel(f"çŠ¶æ€: <b>{current_cat_text}</b>")

            if category_key:
                current_cat_label.setStyleSheet("color: #00A896; font-weight: bold;")
            else:
                current_cat_label.setStyleSheet("color: #FF6F00; font-weight: bold;")

            # 4. åˆ†ç±»æŒ‰é’®
            classify_btn = QPushButton("æ‰‹åŠ¨åˆ†ç±»/ä¿®æ”¹")
            classify_btn.article_id = article_id
            classify_btn.clicked.connect(self._handle_classify_article)

            # å¸ƒå±€æ§ä»¶
            group_layout.addWidget(id_label, 0, 0, 1, 3)
            group_layout.addWidget(text_browser, 1, 0, 1, 3)
            group_layout.addWidget(rec_widget, 2, 0)
            group_layout.addWidget(current_cat_label, 2, 1)
            group_layout.addWidget(classify_btn, 2, 2)

            layout.addWidget(article_group)

        # æ·»åŠ ä¼¸å±•ç©ºé—´ï¼Œç¡®ä¿å†…å®¹é¡¶éƒ¨å¯¹é½
        layout.addStretch()

        # ğŸŒŸ ç­›é€‰çŠ¶æ€æç¤º ğŸŒŸ
        if self.is_filtered:
            self.show_notification(f"å½“å‰æ˜¾ç¤º {len(results)} æ¡æœªåˆ†ç±»æ¡æ–‡ã€‚", is_error=False)

    def _convert_display_key_to_save_key(self, display_key: str) -> str | None:
        """
        å°† L1Name-L2Name-L3Name çš„æ˜¾ç¤ºæ ¼å¼
        è½¬æ¢ä¸º Service è¦æ±‚çš„ L1Key/L2Name/L3Name (ä¾‹å¦‚: 0/èµˆç¾ä¸æ°‘ç”Ÿä¿éšœ/èµˆç¾)
        """
        parts = display_key.split('-')
        if len(parts) != 3:
            return None

        level1_name = parts[0]

        if level1_name == "äº‹åŠ¡ç±»":
            l1_key = "0"
        elif level1_name == "é—®é¢˜ç±»":
            l1_key = "1"
        else:
            return None

        # æ„é€  Service è¦æ±‚çš„ä¿å­˜æ ¼å¼ï¼š L1_KEY/L2_NAME/L3_NAME
        return f"{l1_key}/{parts[1]}/{parts[2]}"

    def _handle_accept_recommendation(self):
        """å¤„ç†ç‚¹å‡»æ¨èåˆ†ç±»æŒ‰é’®çš„äº‹ä»¶ï¼šç›´æ¥é‡‡ç”¨æ¨èåˆ†ç±»å¹¶ä¿å­˜"""
        sender_button = self.sender()
        if not hasattr(sender_button, 'article_id') or not hasattr(sender_button, 'recommendation_key'):
            QMessageBox.critical(self, "é”™è¯¯", "æ— æ³•è¯†åˆ«æŒ‰é’®å…³è”çš„æ¡æ–‡IDæˆ–æ¨èåˆ†ç±»ã€‚")
            return

        article_id = sender_button.article_id
        display_key = sender_button.recommendation_key

        classification_key = self._convert_display_key_to_save_key(display_key)

        if not classification_key:
            QMessageBox.critical(self, "é”™è¯¯", f"æ— æ³•è§£ææ¨èåˆ†ç±»é”® '{display_key}' ä¸ºä¿å­˜æ ¼å¼ï¼Œè¯·æ‰‹åŠ¨åˆ†ç±»ã€‚")
            return

        current_article = next((a for a in self.file_manager.get_batch_articles() if a['article_id'] == article_id),
                               None)
        if not current_article:
            QMessageBox.critical(self, "é”™è¯¯", f"æœªæ‰¾åˆ°æ¡æ–‡ID: {article_id}")
            return

        self._perform_save_classification(article_id, classification_key, current_article)

    def _perform_save_classification(self, article_id, classification_key, current_article):
        """å°†åˆ†ç±»ä¿å­˜åˆ° Serviceï¼Œå¹¶æ›´æ–° UI"""
        try:
            self.file_manager.update_article_classification(article_id, classification_key)

            self.qingshilu_service.save_classification_result(
                current_article['originalText'],
                current_article['analysis'].get('translation', 'N/A'),
                classification_key,
                article_id=article_id
            )

            self.show_notification(f"åˆ†ç±»æˆåŠŸï¼šæ¡æ–‡ {article_id} å·²ä¿å­˜åˆ°: {classification_key}")

            # ğŸŒŸ ä¿®æ­£ï¼šæ ¹æ®å½“å‰çš„ç­›é€‰çŠ¶æ€é‡æ–°æ¸²æŸ“ ğŸŒŸ
            if self.is_filtered:
                all_articles = self.file_manager.get_batch_articles()
                results_to_render = [
                    a for a in all_articles
                    if a.get('classification_key') is None or a.get('classification_key') == ''
                ]
            else:
                results_to_render = self.file_manager.get_batch_articles()

            self._render_batch_results(results_to_render)

        except Exception as e:
            QMessageBox.critical(self, "ä¿å­˜å¤±è´¥", f"åˆ†ç±»ä¿å­˜å¤±è´¥: {e}")

    def _handle_classify_article(self):
        """å¤„ç†æ‰¹é‡ç»“æœä¸­çš„å•æ¡æ¡æ–‡åˆ†ç±»ï¼ˆæ‰‹åŠ¨é€‰æ‹©ï¼‰"""
        sender_button = self.sender()
        if not hasattr(sender_button, 'article_id'):
            QMessageBox.critical(self, "é”™è¯¯", "æ— æ³•è¯†åˆ«æŒ‰é’®å…³è”çš„æ¡æ–‡IDã€‚")
            return

        article_id = sender_button.article_id

        current_article = next((a for a in self.file_manager.get_batch_articles() if a['article_id'] == article_id),
                               None)
        if not current_article:
            QMessageBox.critical(self, "é”™è¯¯", f"æœªæ‰¾åˆ°æ¡æ–‡ID: {article_id}")
            return

        category_dialog = CategorySelectionDialog(self.qingshilu_service.get_category_structure(), self)
        if category_dialog.exec() == QDialog.Accepted:
            classification_key = category_dialog.get_selected_key()

            self._perform_save_classification(article_id, classification_key, current_article)

    def _handle_save_batch_results(self):
        """å¤„ç†æ‰¹é‡ä¿å­˜ç»“æœçš„é€»è¾‘"""
        results = self.file_manager.get_batch_articles()

        if not results:
            QMessageBox.warning(self, "è­¦å‘Š", "æ²¡æœ‰æ‰¹é‡åˆ†æç»“æœå¯ä¾›ä¿å­˜ã€‚")
            return

        file_path, _ = QFileDialog.getSaveFileName(
            self, "ä¿å­˜æ‰¹é‡ç»“æœ", "batch_analysis_results.json", "JSON æ–‡ä»¶ (*.json);;æ‰€æœ‰æ–‡ä»¶ (*)"
        )

        if file_path:
            try:
                with open(file_path, 'w', encoding='utf-8') as f:
                    json.dump(results, f, ensure_ascii=False, indent=4)

                QMessageBox.information(self, "ä¿å­˜æˆåŠŸ", f"æ‰¹é‡ç»“æœå·²æˆåŠŸä¿å­˜åˆ°: {file_path}")
            except Exception as e:
                QMessageBox.critical(self, "ä¿å­˜å¤±è´¥", f"æ— æ³•ä¿å­˜æ–‡ä»¶: {e}")
